### PR TITLE
JAVA-1367: Make protocol negotiation more resilient

### PR DESCRIPTION
**Pull request by adutra**
Motivation:

CASSANDRA-11464 introduced in 3.0.8+, 3.8+ changed logic of protocol
error messages to be sent with the protocol version requested
by the client, instead of a version that the server supports,
to ensure that the client will be able to decode the response properly.

The driver currently uses the returned protocol version to determine
what protocol version to downgrade to and reattempt connection.
Because of this, the driver would retry the same protocol version and
fail again.

Modifications:

This commit changes the renegotiation algorithm to allow
more than one renegotiation attempt. It also makes no more distinction
between user-defined versions and inferred versions.
And finally, it now ignores the server response version
and always retries to connect using attempted version - 1.

Result:

Driver is compatible with Cassandra 3.0.8+, 3.8+.

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/Connection.java [hunk 1]
```diff
@@ -405,9 +405,52 @@ private void incrementAuthErrorMetric() {
         }
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(ProtocolVersion triedVersion, ProtocolVersion serverProtocolVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {} server supports version {}", address, triedVersion, serverProtocolVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion, serverProtocolVersion);
+    private static final Pattern SUPPORTED_VERSIONS_PATTERN = Pattern.compile("supported versions are \\(([^)]+?)\\)");
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Connection.java]: Since C* has been pretty inconsistent about the error messages it returns by release, what about just not parsing the error message at all and then tying the next version to try based on the protocol version that we previously attempted?   The problem we are experiencing at the moment is because we were using information provided by the server, so I think we should just not trust it unless we can get something in the protocol specification about how to handle this scenario.

In the general case, we could simply just retry at the attempted protocol version - 1.  The only downside of this is that it may take multiple attempts to find the right protocol version, but this is only done at initialization anyways, so it doesn't seem to have that much cost.  We could also add a method to `ProtocolVersion` for requesting the next version to try in the case where it's more complex than that, this is actually what the nodejs-driver [is proposing](https://github.com/datastax/nodejs-driver/blob/1ed4a37dc010b0e57e0ae55bfb4704d97e2d622d/lib/types/protocol-version.js#L113-L127), although it still keeps the optimization to use the protocol version in the error response if it is different (like we also do here).

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Connection.java]: > Since C* has been pretty inconsistent about the error messages it returns

Not so much actually, I crawled through the revisions of `Frame.java` and found 4 different messages only.

> In the general case, we could simply just retry at the attempted protocol version - 1

I guess what you mean is _if_ the server version differs from the client one, then use that version, otherwise downgrade to N-1, is that right? The only problem I see with the "-1" approach is that downgrading when the server replies with the same version is not useful at all (any lower version will fail as well – what we need in that case is actually upgrade).

Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/Connection.java]: I don't understand, the client starts negotiation with its highest version. If it started with vN, it doesn't support N+1 nor N+2.

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Connection.java]: hmm now that I think of it, maybe we could simplify the algorithm to:

1. if the server version differs from the client one, then use that version;
2. otherwise upgrade to server version + 1, if possible;
3. otherwise fail.

Step 2 sounds counterintuitive, but that could actually work for all cases. Will give some more thought.

Comment (tolbertam) [reply driver-core/src/main/java/com/datastax/driver/core/Connection.java]: > I guess what you mean is if the server version differs from the client one, then use that version, otherwise downgrade to N-1, is that right? The only problem I see with the "-1" approach is that downgrading when the server replies with the same version is not useful at all (any lower version will fail as well – what we need in that case is actually upgrade).

Unless i'm misunderstanding, we would never upgrade the protocol version, only downgrade.

I think the most straightforward approach is what @jorgebay came up with in datastax/nodejs-driver#204.  That implementation is similar to what you said with a few caveats:

    1. if the server version differs from the client one, then use that version;
    2. otherwise downgrade from the previous protocol version you tried

Steps 1 and 2 would then be repeated until a suitable protocol version is found.

I think all the other drivers (other than C# and nodejs) only do step 2.

'downgrade' could mean different things depending on the context of the protocol version, so would be nice to have a method on `ProtocolVersion` for asking for the lower protocol version.

**Comment by adutra** [comment]
Addressed all feedback so far:

1. Removed all parsing of error messages;
2. Simplified renegotiation algorithm.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 2]
```diff
@@ -1504,18 +1504,7 @@ long delayMs() {
...
-                    logger.debug("Cannot connect with protocol {}, trying {}", e.getUnsupportedVersion(), e.getServerVersion());
-
-                    connectionFactory.protocolVersion = e.getServerVersion();
-                    try {
-                        controlConnection.connect();
-                    } catch (UnsupportedProtocolVersionException e1) {
-                        throw new DriverInternalError("Cannot connect to node with its own version, this makes no sense", e);
-                    }
-                }
+                connect();
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: I like that a lot of logic was moved into separate methods, makes code a lot easier to follow 👍

File: driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+import com.datastax.driver.core.exceptions.NoHostAvailableException;
+import com.datastax.driver.core.exceptions.UnsupportedProtocolVersionException;
+import org.testng.annotations.Test;
+
+import static com.datastax.driver.core.ProtocolVersion.*;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@CreateCCM(CreateCCM.TestMode.PER_METHOD)
+public class ProtocolVersionRenegotiationTest extends CCMTestsSupport {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java:25]: Excellent tests!

**Comment by tolbertam** [comment]
So far this looks really great.  Going to spend some time experimenting with it, but the test coverage looks like it covers the scenarios well.

File: driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        Cluster cluster = connectWithoutVersion();
+        assertThat(actualProtocolVersion(cluster)).isEqualTo(V1);
+    }
+
+    /**
+     * @jira_ticket JAVA-1367
+     */
+    @Test(groups = "short")
+    @CCMConfig(version = "2.0.17", createCluster = false)
+    public void should_renegotiate_when_renegotiation_enabled_and_initial_version_too_high_2_0_17() throws Exception {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: This is more of a note for me.  But I realized that in our CI environment we choose the JDK to run C* against based on the global cassandra version for the test.  If C* is less than 2.1, we choose JDK 7 because C* 2.0 doesn't work with JDK 8.  However, I think this assertion is untrue.  I think the issue is that C* 2.0 doesn't compile on Java 8 ([CASSANDRA-7936](https://issues.apache.org/jira/browse/CASSANDRA-7936)), but it will still run on it (works for me locally at least).  I'll fix that.

Comment (tolbertam) [reply driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: The other interesting thing about these tests is they run no matter what `-Dcassandra.version` is and I think that is ok.  it is good to get as much validation as possible done in one run.

**Review by tolbertam — APPROVED** [review]
Have experimented with this locally a bit and it's working great 👍

File: driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java (hunk 4 above)
Comment (adutra) [reply driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: `@CCMConfig(version = "XXX")` forces a specific version instead of skipping the test if the version doesn't match the default one. I think it is required sometimes too, as tests like these could fail if run against any other version.

File: driver-core/src/main/java/com/datastax/driver/core/ProtocolVersion.java [hunk 5]
```diff
@@ -73,6 +68,26 @@ public int toInt() {
...
+                return V4;
+            case V4:
+                return V3;
+            case V3:
+                return V2;
+            case V2:
+                return V1;
+        }
+        return null;
+    }
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/ProtocolVersion.java]: How about
```
return INT_TO_VERSION.get(asInt - 1);
```

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 6]
```diff
@@ -1593,10 +1582,35 @@ else if (host.state == Host.State.DOWN)
...
 
+        private void connect() {
+            try {
+                controlConnection.connect();
+            } catch (UnsupportedProtocolVersionException e) {
+                ProtocolVersion newVersion = e.getServerVersion();
+                if (newVersion == e.getUnsupportedVersion()) {
+                    newVersion = newVersion.getLowerSupported();
+                    if (newVersion == null) {
+                        throw e;
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: This was a bit hard to follow for me. I'm pushing a proposal with intermediate variables and comments.

**Review by olim7t — APPROVED** [review]
lgtm with 2 minor comments.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 7]
```diff
@@ -1593,15 +1593,15 @@ private void connect() {
                 if (serverResponseVersion == attemptedVersion) {
                     // We're post CASSANDRA-11464: the server replies with the same version that the client tried (to
                     // ensure that the client can decode the response).
-                    // Retry with a lower version (if it's too low we'll eventually reach our min and fail).
+                    // Retry with a lower version (if it's too low we'll fail at the next attempt, as we only retry once).
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: Right, I missed that.
So what happens if the server supports up to v10 and the client starts with v12?

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: > what happens if the server supports up to v10 and the client starts with v12?

The server would reply with v10. If the driver also supports v10, then v10 will be used; otherwise, it will fail when decoding the response frame ([here](https://github.com/datastax/java-driver/blob/3.x/driver-core/src/main/java/com/datastax/driver/core/Frame.java#L93-L93) more precisely).

Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: OK, so the last piece I was missing is that C* only replies with the same version if the driver started too low (e.g. C* supports v12+ and the driver started with v10).
In that case, what's the point of even retrying? If the driver was already too low, going lower won't help.

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: > In that case, what's the point of even retrying?

That was my point since the beginning. But @jorgebay convinced me: the fact that retrying at N-1 is necessarily going to fail is not a documented feature, but an implementation detail at this point. We don't know if this is going to always be the case, so retrying at N-1 is more future-proof.

Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: > the fact that retrying at N-1 is necessarily going to fail is not a documented feature

If we're going that route, the same could be said for N-2.

This is getting complicated. How about going back to the simple algorithm used [before JAVA-361](https://github.com/datastax/java-driver/blob/2.1.0/driver-core/src/main/java/com/datastax/driver/core/Cluster.java#L1143-L1168):
* try with version N
* if we get a PROTOCOL_ERROR, simply retry with the driver's previous version. Ignore the version it's encoded in (that's not documented in the spec either after all).
* repeat until we run out of driver's supported versions.

Yes, we might do a couple of extra retries in cases where there is no chance of success. It's not a big deal, negotiation is not supposed to be efficient anyway, and we discourage its use in production.

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: > This is getting complicated. How about going back to the simple algorithm used before JAVA-361.

I don't have anything against the idea. I will push the modifications shortly if nobody disagrees.

Comment (adutra) [reply driver-core/src/main/java/com/datastax/driver/core/Cluster.java]: Sorry, but your suggestion makes some tests fail. The main reason is that we do not retry more than once, so if we start too high, the second attempt will still be too high and fail, whereas currently the second attempt could potentially be much lower that the initial one. 

On the light of this, I'm actually reconsidering my position: I think we should rather keep the current algorithm.

**Comment by adutra** [comment]
As discussed, I'm proposing a revised algorithm that basically gives us pre-ProtocolVersion enum behavior.

**Comment by tolbertam** [comment]
🤦‍♂️ I spelled 'accommodate' wrong in my commit message for `f228596` when you squash can you either fix that or not include that in the final commit?

**Comment by tolbertam** [comment]
Changes look good to me 👍

**Comment by adutra** [comment]
Squashed and rebased. Ready to be merged.

File: driver-core/src/main/java/com/datastax/driver/core/Cluster.java [hunk 8]
```diff
@@ -1582,6 +1583,30 @@ else if (host.state == Host.State.DOWN)
...
+                    }
+                    // Do not trust version of server's response, as C* behavior in case of protocol negotiation is not
+                    // properly documented, and varies over time (specially after CASSANDRA-11464). Instead, always
+                    // retry at attempted version - 1, if such a version exists; and otherwise, stop and fail.
+                    ProtocolVersion attemptedVersion = e.getUnsupportedVersion();
+                    ProtocolVersion retryVersion = attemptedVersion.getLowerSupported();
+                    if (retryVersion == null) {
+                        throw e;
+                    }
+                    logger.info("Cannot connect with protocol version {}, trying with {}", attemptedVersion, retryVersion);
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/Cluster.java:1604]: I raised the level of this message from DEBUG.

File: driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java [hunk 9]
```diff
@@ -53,45 +62,31 @@ public void should_fail_when_renegotiation_disabled_3_10() throws Exception {
...
     }
 
     /**
      * @jira_ticket JAVA-1367
      */
     @Test(groups = "short")
-    @CCMConfig(version = "2.0.17", createCluster = false)
-    public void should_renegotiate_and_initial_version_too_high_2_0_17() throws Exception {
-        Cluster cluster = connectWithVersion(V4);
-        assertThat(actualProtocolVersion(cluster)).isEqualTo(V2);
```
Comment (olim7t) [inline driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: There should be no negotiation when an explicit version is provided. `withProtocolVersion`'s javadocs say:
> This method allows to force the use of a particular protocol version.

Comment (tolbertam) [reply driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: I think that might have got lost in the shuffle or there might have been a misinterpretation as the commit message includes:

>  It also makes no more distinction between user-defined versions and inferred versions.

Agreed that renegotiation should only occur if the protocol verison is not provided.

Comment (adutra) [reply driver-core/src/test/java/com/datastax/driver/core/ProtocolVersionRenegotiationTest.java]: Sorry, that wasn't at all what I understood from our conversation.
